ops/pebble.py

```diff
@@ -1966,3 +1952,162 @@ def send_signal(self, sig: typing.Union[int, str], services: typing.List[str]):
+                    self._headers = None
+                else:
+                    # Insufficient data to continue
+                    break
+            elif self._part_type == 'files':
+                next_boundary_index = self._get_next_boundary_index()
+                if next_boundary_index is not None:
+                    # Next boundary's location is known; this is the final write.
+                    data = self._buffer[:next_boundary_index]
+                    with self._get_open_tempfile() as outfile:
+                        outfile.write(data)
+                    # Advance the buffer to point at the next boundary
+                    self._buffer = self._buffer[next_boundary_index:]
+                    self._headers = None
+                else:
+                    # Data present but boundary not yet found.
+                    # Write data from buffer, but don't accidentally write data which may belong
+                    # to a boundary line.
+                    safe_bound = len(self._buffer) - len(self._terminal_boundary)
+                    data = self._buffer[:safe_bound]
+
+                    # Don't open/close file for intermediate writes
+                    outfile = self._get_open_tempfile()
+                    outfile.write(data)
+
+                    # Advance the buffer for bytes consumed
+                    self._buffer = self._buffer[safe_bound:]
+
+                    # Insufficient data to continue
+                    break
+
+    def _parse_headers(self, data):
+        parser = email.parser.BytesFeedParser()
+        parser.feed(data)
+        message = parser.close()
+        return message
+
+    def _detect_invalid_content_disposition(self):
+        content_disposition = self._headers.get_content_disposition()
+        if content_disposition != 'form-data':
+            raise RuntimeError(
+                'unexpected content disposition: {}'.format(content_disposition))
+
+    def _prepare_for_part_body(self):
+        name = self._headers.get_param('name', header='content-disposition')
+        if name == 'response':
+            part_type = 'response'
+        elif name == 'files':
+            filename = self._headers.get_filename()
+            if filename is None:
+                raise NotImplementedError('multipart "files" part missing filename')
+            self._prepare_tempfile(filename)
+            part_type = 'files'
+        else:
+            raise RuntimeError(
+                'unexpected name in content-disposition header: {}'.format(name))
+        return part_type
+
+    def _get_next_boundary_index(self) -> typing.Optional[int]:
+        indices = set()
+        for boundary in (self._boundary, self._terminal_boundary):
+            try:
+                index = self._buffer.index(boundary)
+            except ValueError:
+                pass
+            else:
+                indices.add(index)
+        return min(indices) if indices else None
```

Comment:
It seems heavy to use a for loop and a set for this. Also, you can use `buffer.find()` instead of `buffer.index()` to avoid the exception. Also, it seems to me the terminal boundary should never be before the regular boundary, so we could check that. What about something like this:

```
def _get_next_boundary_index(self):
    boundary_index = self._buffer.find(self._boundary)
    terminal_index = self._buffer.find(self._terminal_boudnary)
    if boundary_index >= 0:
        if terminal_index >= 0 and terminal_index < boundary_index:
            raise RuntimeError('terminal boundary found before regular boundary')
        return boundary_index
    if terminal_index >= 0:
        return terminal_index
    return None
```

Reply:
Looks like this was done in a later commit :)